### PR TITLE
GraphQl

### PR DESCRIPTION
**Pull request by soyuka**

### CODE REVIEW

File: src/Core/GraphQl/Serializer/SerializerContextBuilder.php [hunk 1]
```diff
@@ -28,38 +29,45 @@
...
 
     public function create(?string $resourceClass, string $operationName, array $resolverContext, bool $normalization): array
     {
-        $resourceMetadata = $resourceClass ? $this->resourceMetadataFactory->create($resourceClass) : null;
-
-        $context = [
-            'resource_class' => $resourceClass,
-            'graphql_operation_name' => $operationName,
-        ];
+        $context = ['operation_name' => $operationName, 'graphql_operation_name' => $operationName];
```
Comment (alanpoulain) [inline src/Core/GraphQl/Serializer/SerializerContextBuilder.php:43]: Why setting `operation_name`?

File: src/Core/GraphQl/Serializer/SerializerContextBuilder.php [hunk 2]
```diff
@@ -28,38 +29,45 @@
...
         }
 
-        if ($resourceMetadata) {
-            $context['input'] = $resourceMetadata->getGraphqlAttribute($operationName, 'input', null, true);
-            $context['output'] = $resourceMetadata->getGraphqlAttribute($operationName, 'output', null, true);
+        if ($operation) {
+            $context['input'] = $operation->getInput();
+            $context['output'] = $operation->getOutput();
 
             $key = $normalization ? 'normalization_context' : 'denormalization_context';
```
Comment (alanpoulain) [inline src/Core/GraphQl/Serializer/SerializerContextBuilder.php:65]: `$key` is not used anymore.

File: src/Core/GraphQl/Subscription/SubscriptionManager.php [hunk 3]
```diff
@@ -92,7 +92,7 @@ public function getPushPayloads($object): array
         foreach ($subscriptions as [$subscriptionId, $subscriptionFields, $subscriptionResult]) {
             $resolverContext = ['fields' => $subscriptionFields, 'is_collection' => false, 'is_mutation' => false, 'is_subscription' => true];
 
-            $data = ($this->serializeStage)($object, $resourceClass, 'update', $resolverContext);
+            $data = ($this->serializeStage)($object, $resourceClass, 'update_subscription', $resolverContext);
```
Comment (alanpoulain) [inline src/Core/GraphQl/Subscription/SubscriptionManager.php:95]: Why?

File: src/Core/GraphQl/Type/FieldsBuilder.php [hunk 4]
```diff
@@ -131,51 +134,56 @@ public function getCollectionQueryFields(string $resourceClass, ResourceMetadata
...
+            $fieldConfiguration['args'] += ['input' => $this->getResourceFieldConfiguration(null, null, $deprecationReason, $resourceType, $resourceClass, true, $subscriptionName)];
         }
 
         if (!$fieldConfiguration) {
             return [];
         }
 
-        $subscriptionFields[$subscriptionName.$resourceMetadata->getShortName().'Subscribe'] = $fieldConfiguration;
+        // TODO: 3.0 change this
+        if ('update_subscription' === $subscriptionName) {
```
Comment (alanpoulain) [inline src/Core/GraphQl/Type/FieldsBuilder.php:174]: Why is it needed to do this?

File: src/Core/GraphQl/Type/FieldsBuilder.php [hunk 5]
```diff
@@ -209,30 +217,35 @@ public function getResourceObjectTypeFields(?string $resourceClass, ResourceMeta
...
         }
 
         ++$depth; // increment the depth for the call to getResourceFieldConfiguration.
 
         if (null !== $resourceClass) {
             foreach ($this->propertyNameCollectionFactory->create($resourceClass) as $property) {
-                $propertyMetadata = $this->propertyMetadataFactory->create($resourceClass, $property, ['graphql_operation_name' => $subscriptionName ?? $mutationName ?? $queryName]);
+                $context = [
+                    'normalization_groups' => $operation->getNormalizationContext()['groups'] ?? null,
+                    'denormalization_groups' => $operation->getDenormalizationContext()['groups'] ?? null,
```
Comment (alanpoulain) [inline src/Core/GraphQl/Type/FieldsBuilder.php:230]: The groups were not used before, why this change?

File: src/Core/GraphQl/Type/FieldsBuilder.php [hunk 6]
```diff
@@ -495,12 +519,20 @@ private function convertType(Type $type, bool $input, ?string $queryName, ?strin
...
+        $resourceClass = '' === $resourceClass ? $rootResource : $resourceClass;
+        if ($resourceClass) {
+            $resourceMetadataCollection = $this->resourceMetadataCollectionFactory->create($resourceClass);
+            try {
+                $operation = $resourceMetadataCollection->getGraphQlOperation($operationName);
+            } catch (OperationNotFoundException $e) {
+            }
+        }
+
+        return $forceNullable || !$graphqlType instanceof NullableType || $type->isNullable() || ($operation instanceof Mutation || $operation instanceof Subscription)
```
Comment (alanpoulain) [inline src/Core/GraphQl/Type/FieldsBuilder.php:535]: This is not the same as before. The type is nullable only when it's an `update` mutation.

File: src/Core/GraphQl/Serializer/SerializerContextBuilder.php (hunk 1 above)
Comment (soyuka) [reply src/Core/GraphQl/Serializer/SerializerContextBuilder.php:43]: Because it may be used by the IRI Converter.

File: src/Core/GraphQl/Subscription/SubscriptionManager.php (hunk 3 above)
Comment (soyuka) [reply src/Core/GraphQl/Subscription/SubscriptionManager.php:95]: In the resource metadata I'm creating the Subscription operation, having that name.

File: src/Core/GraphQl/Type/FieldsBuilder.php (hunk 4 above)
Comment (soyuka) [reply src/Core/GraphQl/Type/FieldsBuilder.php:174]: I don't want to break any tests, therefore names must stay the same. The `update` operation already exists and is a Mutation we need unique operation names.

File: src/Core/GraphQl/Type/FieldsBuilder.php (hunk 5 above)
Comment (soyuka) [reply src/Core/GraphQl/Type/FieldsBuilder.php:230]: This is a complex issue, I want ResourceMetadataCollectionFactory to depend on PropertyMetadataFactory so that we can compute identifiers. The SerializerPropertyMetadataFactory also relies on ResourceMetadataFactories to compute groups which leads to a recursive dependency. I'm still trying to look for a solution, getting the groups before calling the SerializerPropertyMetadataFactory seems to be the less complex thing to do.

Comment (alanpoulain) [reply src/Core/GraphQl/Type/FieldsBuilder.php:230]: I think it can create issues in nested type construction, but I'm not really sure. It needs to be checked.
